### PR TITLE
Add plumbing to support upgrade testing. 

### PR DESCRIPTION
**Pull request by rramkumar1**
Adds the ability to run a test inside a sandbox continuously. This allows us to test functionality while the ingress is being updated.

### CODE REVIEW

File: cmd/e2e-test/basic_upgrade_test.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+				AddPath("test.com", "/", "service-1", port80).
+				Build(),
+			numForwardingRules: 1,
+			numBackendServices: 2,
+		},
+	} {
+		expectUnreachable := true
+		tc := tc // Capture tc as we are running this in parallel.
+		Framework.RunContinuouslyWithSandbox(tc.desc, t, func(t *testing.T, s *e2e.Sandbox) {
+			if s.IsFresh {
```
Comment (rramkumar1) [inline cmd/e2e-test/basic_upgrade_test.go]: Not sure if I want to put the notion of "fresh" inside each sandbox or have the framework keep track of each sandbox's freshness

File: pkg/e2e/helpers.go [hunk 2]
```diff
@@ -39,7 +39,7 @@ const (
 )
 
 // WaitForIngress to stabilize.
-func WaitForIngress(s *Sandbox, ing *v1beta1.Ingress) (*v1beta1.Ingress, error) {
+func WaitForIngress(s *Sandbox, ing *v1beta1.Ingress, expectUnreachable bool) (*v1beta1.Ingress, error) {
```
Comment (MrHohn) [inline pkg/e2e/helpers.go]: Probably create a `WaitOptions` struct?

File: pkg/e2e/helpers.go [hunk 3]
```diff
@@ -53,8 +53,13 @@ func WaitForIngress(s *Sandbox, ing *v1beta1.Ingress) (*v1beta1.Ingress, error)
 		result := validator.Check(context.Background())
 		if result.Err == nil {
 			return true, nil
+		} else {
+			if expectUnreachable {
+				return false, nil
+			} else {
```
Comment (MrHohn) [inline pkg/e2e/helpers.go]: Unnecessary else block?

File: pkg/e2e/helpers.go [hunk 4]
```diff
@@ -53,8 +53,13 @@ func WaitForIngress(s *Sandbox, ing *v1beta1.Ingress) (*v1beta1.Ingress, error)
...
+				return true, fmt.Errorf("Unexpected error from validation: %v", result.Err)
```
Comment (MrHohn) [inline pkg/e2e/helpers.go]: return false, fmt.Errorf("Unexpected error from validation: %v", result.Err)? `Poll()` will return once there is an error.

File: cmd/e2e-test/basic_upgrade_test.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+		{
+			desc: "http one path",
+			ing: fuzz.NewIngressBuilder("", "ingress-1", "").
+				AddPath("test.com", "/", "service-1", port80).
+				Build(),
+			numForwardingRules: 1,
+			numBackendServices: 2,
+		},
+	} {
+		expectUnreachable := true
```
Comment (MrHohn) [inline cmd/e2e-test/basic_upgrade_test.go]: Why do we need `expectUnreachable=true` at the first run?

File: cmd/e2e-test/security_policy_test.go [hunk 6]
```diff
@@ -111,7 +111,7 @@ func TestSecurityPolicyEnable(t *testing.T) {
 
 		t.Logf("Checking on relevant backend service whether security policy is properly attached")
 
-		testIng, err = e2e.WaitForIngress(s, testIng)
+		testIng, err = e2e.WaitForIngress(s, testIng, true)
```
Comment (MrHohn) [inline cmd/e2e-test/security_policy_test.go]: Humm..Why should it expect unreachable?

File: cmd/e2e-test/basic_upgrade_test.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+	"github.com/kr/pretty"
+	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/ingress-gce/pkg/e2e"
+	"k8s.io/ingress-gce/pkg/fuzz"
+	"k8s.io/ingress-gce/pkg/fuzz/features"
+)
+
+func TestBasicUpgrade(t *testing.T) {
```
Comment (MrHohn) [inline cmd/e2e-test/basic_upgrade_test.go]: Will this upgrade test run with the other tests or will we have a explicit upgrade job? How would we trigger the upgrade action?

Comment (rramkumar1) [reply cmd/e2e-test/basic_upgrade_test.go]: If you look in run.sh, I added a couple line to trigger just the upgrade tests if we want.

File: cmd/e2e-test/basic_upgrade_test.go (hunk 5 above)
Comment (rramkumar1) [reply cmd/e2e-test/basic_upgrade_test.go]: In the first run, we will experience the programming latency for the LB so we expect that we will get back some 404's/502's.

File: cmd/e2e-test/security_policy_test.go (hunk 6 above)
Comment (rramkumar1) [reply cmd/e2e-test/security_policy_test.go]: See above comment.

File: pkg/e2e/helpers.go (hunk 4 above)
Comment (rramkumar1) [reply pkg/e2e/helpers.go]: I want it to return in this case. If expectUnreachable is false, meaning we expect no 404's/502's, then I want to return the error if there was one.

File: pkg/e2e/helpers.go (hunk 3 above)
Comment (rramkumar1) [reply pkg/e2e/helpers.go]: Done

File: pkg/e2e/helpers.go (hunk 2 above)
Comment (rramkumar1) [reply pkg/e2e/helpers.go]: Done

File: cmd/e2e-test/basic_upgrade_test.go (hunk 7 above)
Comment (MrHohn) [reply cmd/e2e-test/basic_upgrade_test.go]: Thanks, should have seen that.

File: pkg/e2e/helpers.go (hunk 4 above)
Comment (MrHohn) [reply pkg/e2e/helpers.go]: I see, that makes sense. I was mistakenly thinking expectUnreachable means we expect it to be always unreachable.

**Comment by rramkumar1** [comment]
Discussed offline and will go with a different approach.

/close.

**Pull request closed without merging**